Add `http_parser` library dependency

staticfloat opened the pull request:
Tested to be building `libhttp_parser` and using it with `libgit2` on OSX and Linux.

From what I can tell of the `libgit2` build system, it will not use a system-provided `http_parser` on Windows, but this PR will build it anyway.

wildart committed:
added libhttp_parser build script

staticfloat committed:
Add http_parser library dependency

Includes cleaned-up patches for http_parser Makefile

tkelman commented on deps/dsfmt.mk:
```diff
@@ -43,8 +43,7 @@ define DSFMT_INSTALL
 	dsfmt,dsfmt-$(DSFMT_VER), \
-	DSFMT_INSTALL,, \
-	$$(DSFMT_OBJ_TARGET), \
+	DSFMT_INSTALL,,, \
```
looks unrelated?

staticfloat replied:
Yeah, this is a separate commit.  I'll move it out for now.

tkelman commented on deps/http_parser.mk:
```diff
@@ -0,0 +1,39 @@
+configure-http_parser: extract-http_parser
+compile-http_parser: $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-compiled
+fastcheck-http_parser: check-http_parser
+check-http_parser: compile-http_parser
```
does it have unit tests?

staticfloat replied:
Heh, I went through the trouble of adding the `$(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/build-checked` recipe, but then didn't make it a target of anything.  :P

tkelman commented on deps/http_parser.version:
```diff
@@ -0,0 +1,2 @@
+HTTPPARSER_BRANCH=master
```
use a release tag

tkelman commented on deps/http_parser.mk:
```diff
@@ -0,0 +1,39 @@
+	$(INSTALL_NAME_CMD)libhttp_parser.$(SHLIB_EXT) $(build_shlibdir)/libhttp_parser.$(SHLIB_EXT)))
+
+$(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/http_parser-makefile.patch-applied: $(BUILDDIR)/$(HTTPPARSER_SRC_DIR)/source-extracted
+	cd $(BUILDDIR)/$(HTTPPARSER_SRC_DIR) && patch -p1 -f < $(SRCDIR)/patches/http_parser-makefile.patch
```
patching git-external deps ends up causing issues. we should upstream any relevant parts of this, or set make flags for others

staticfloat replied:
Upstream has had something similar to this patch open as a [PR since June](https://github.com/nodejs/http-parser/pull/321), it looks like upstream is a little slow to respond to PRs.  The patch carried here is based off of the patch in that PR, with some extra stuff for Julia.  I'll see if I can pare the patch down, I was just copying it wholesale from wildart's original work.

tkelman commented on deps/patches/http_parser-makefile.patch:
```diff
@@ -0,0 +1,112 @@
++	rm $(DESTDIR)$(INCLUDEDIR)/http_parser.h
++	rm $(DESTDIR)$(LIBDIR)/$(SONAME)
++	rm $(DEstdiR)$(LIBDIR)/libhttp_parser.$(SOEXT)
++	rm $(DEstdiR)$(LIBDIR)/libhttp_parser.a
```
what's with the capitalization here?

staticfloat replied:
This was part of the original patch in the main http-parser PR, I'll clean this patch up.

tkelman commented on deps/patches/http_parser-makefile.patch:
```diff
@@ -0,0 +1,112 @@
+ 		http_parser.tar libhttp_parser.so.* \
+ 		url_parser url_parser_g parsertrace parsertrace_g \
+-		*.exe *.exe.so
++		*.exe.so *.dll
```
exe should probably still be cleaned

staticfloat replied:
Agreed.

tkelman reviewed (changes requested):
should also go in license and readme

how hard is it to make libgit2 use an external http-parser library on windows?

staticfloat commented:
Alright, I think I've addressed all comments so far.  I have no idea what it takes to make libgit2 use a system-provided `http_parser`, looking at the [cmake file](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L238-L241), it looks like if we use `WinHTTP` as the transport layer, we have no choice but to build the libgit2-provided `http_parser`.  As [this is the default](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L83) and we don't seem to override that anywhere in our building of `libgit2`, I'd say it's currently impossible to do so.

I don't know what the advantages/disadvantages of using `WinHTTP` are over "the built-in HTTP transport", so I don't know if changing that libgit2 option is a good idea or not.

tkelman commented on deps/Makefile:
```diff
@@ -22,7 +22,7 @@ include $(SRCDIR)/tools/git-external.mk
 # custom Makefile rules: openlibm dsfmt suitesparse-wrapper suitesparse lapack openblas utf8proc objconv osxunwind
 # entirely custom: virtualenv
-# CMake libs: libgit2 libssh2 mbedtls
+# CMake libs: http_parser libgit2 libssh2 mbedtls
```
it doesnt use cmake...

staticfloat replied:
doh

tkelman commented on deps/http_parser.mk:
```diff
@@ -0,0 +1,39 @@
+# Also manually set PREFIX to DESTDIR
+HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix)
+
+$(eval $(call staged-install, \
```
does this create the manifest and uninstall targets?

staticfloat replied:
[Yes](https://github.com/JuliaLang/julia/blob/548465b6ded7d2d4f908760e66329d7fcdfd356d/deps/tools/common.mk#L159-L163) and [Yes](https://github.com/JuliaLang/julia/blob/548465b6ded7d2d4f908760e66329d7fcdfd356d/deps/tools/common.mk#L132-L134)

staticfloat committed:
Properly identify `http_parser`'s build system in the doc comments

staticfloat committed:
Install `libhttp_parser` when running `make install`

Another bug pre-emptively squashed through proper application of CI

staticfloat committed:
Add `http_parser` documentation into `LICENSE.md` and `README.md`

staticfloat commented:
@tkelman is this all good to go from your viewpoint?

tkelman commented:
Has libgit2 modified their copy of http-parser to work with winhttp?

Should also trial run this through the buildbots first.

staticfloat commented:
> Has libgit2 modified their copy of http-parser to work with winhttp?

It looks to me like [winhttp is on by default](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L80-L84), so I'd say yes.

> Should also trial run this through the buildbots first.

[macOS build](https://build.julialang.org/builders/package_osx10.9-x64/builds/1015): Build finished, `libhttp_parser.2.7.1.dylib` confirmed to be in the archive.
[linux32 build](https://build.julialang.org/builders/package_tarball32/builds/998): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[linux64 build](https://build.julialang.org/builders/package_tarball64/builds/1032): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[arm build](https://build.julialang.org/builders/package_tarballarm/builds/974): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[ppc64le build](https://build.julialang.org/builders/package_tarballppc64le/builds/234): Build finished, `libhttp_parser.so.2.7.1` confirmed to be in the archive.
[win32 build](https://build.julialang.org/builders/package_win6.2-x86/builds/887): Build finished, I can see that `libhttp_parser.2.7.1.dll` has been built, and I assume it is also in the archive, but I can't see that just from the logs.
[win64 build](https://build.julialang.org/builders/package_win6.2-x64/builds/958): Build finished, I can see that `libhttp_parser.2.7.1.dll` has been built, and I assume it is also in the archive, but I can't see that just from the logs.

tkelman commented:
I meant the source code. If it's a matter of adding a define when building http-parser, we can easily do that ourselves.

tkelman commented on deps/patches/http_parser-makefile.patch:
```diff
@@ -0,0 +1,98 @@
+ SONAME ?= libhttp_parser.2.7.1.dylib
+ SOEXT ?= dylib
++else ifeq (winnt,$(PLATFORM))
++SONAME ?= libhttp_parser-2.7.dll
```
why was this not the soname on the windows buildbot?

staticfloat replied:
Errant brain interpolation

staticfloat replied:
Edited and fixed above

staticfloat commented:
> I meant the source code. If it's a matter of adding a define when building http-parser, we can easily do that ourselves.

If you mean you want to normalize across platforms and _not_ use WinHTTP, then I think all we have to do is to add `-DWINHTTP=OFF` to [this list of libgit2 CMake options](https://github.com/JuliaLang/julia/blob/f7a611eb1ab00d4eef9a50a3a1f09fc1a8ded6e3/deps/libgit2.mk#L27).  I just don't know what the benefit of doing/not doing that is, so I stayed away from it.

tkelman commented:
winhttp means the windows native crypto backend. We want to use that for https support (we could try using mbedtls instead but that's still fairly experimental, best to keep it linux only). That shouldn't make a huge difference to whether libgit2 builds its vendored copy of http-parser or links to an external copy, unless they've made source changes that are not upstreamed.

staticfloat commented:
Great, then I think we're ready to go.  As it stands right now, `http_parser` is always built separately (so it's still available for use on Windows), but `libgit2` does not use it on Windows, since WinHTTP is enabled by default.

tkelman commented:
>  if we use WinHTTP as the transport layer, we have no choice but to build the libgit2-provided http_parser

does libgit2 use http-parser on windows or not? if yes, then we should link against the standalone copy

staticfloat commented:
Ah, I finally understand what you're asking.  It doesn't look like they have support for using _external_ http-parser on windows.  They [force the usage](https://github.com/libgit2/libgit2/blob/master/CMakeLists.txt#L239-L241) of their internal copy, but [their internal copy](https://github.com/libgit2/libgit2/blob/master/deps/http-parser/http_parser.c) has no mention of windows `#ifdef`'s or whatnot, and doesn't use the `GIT_WINHTTP` preprocessor macro that is defined when WinHTTP is used, so I theorize we should be able to link in our own http-parser if we wish.  We just need to mess around with the CMakeLists file a bit.

staticfloat commented:
I've made a PR to upstream [here](https://github.com/libgit2/libgit2/pull/4015), we'll see what they think of it.

quinnj commented:
Bump, are we waiting on something upstream?

tkelman commented:
Cleaning the PR up and/or incorporating the patch here

staticfloat committed:
Add libgit2 patch

staticfloat commented:
I've added the patch here so we can at least ensure that it works nicely for us.  I'm running it through the buildbots, here are the builds, I'll post links to the binaries so that users such as @quinnj can ensure that they work:

[OSX build](https://build.julialang.org/builders/package_osx10.9-x64/builds/1059): Pending
[Linux x86 build](https://build.julialang.org/builders/package_tarball32/builds/1043): Pending
[Linux x64 build](https://build.julialang.org/builders/package_tarball64/builds/1074): Pending
[Linux ARM build](https://build.julialang.org/builders/package_tarballarm/builds/1016): Pending
[Linux PPC64le build](https://build.julialang.org/builders/package_tarballppc64le/builds/276): Pending
[Windows x64 build](https://build.julialang.org/builders/package_win6.2-x64/builds/1000): Pending
[Windows x86 build](https://build.julialang.org/builders/package_win6.2-x86/builds/933): Pending

staticfloat committed:
Explicitly set `LIBDIR` for http_parser

This ensures that `.dll` files end up in `<prefix>/bin`

staticfloat commented:
The `.dll` file isn't getting put in the right place on Windows; it's being put into `/lib` instead of `/bin`.  I've pushing the latest commit (d15edf2) through the windows builders, we'll see if this fixes things up.  In the meantime, the earlier commit worked on OSX and Linux 64:

```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser"))
"/tmp/julia-121f07d151/bin/../lib/julia/libhttp_parser.so.2.7.1"
```

```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser"))
"/Users/sabae/Desktop/Julia-0.6.app/Contents/Resources/julia/lib/julia//libhttp_parser.dylib"
```

staticfloat commented:
New binaries available:

Win32 [download link](https://s3.amazonaws.com/julianightlies/bin/winnt/x86/0.6/julia-0.6.0-d15edf2f3b-win32.exe).
Win64 [download link](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-d15edf2f3b-win64.exe).

Things are not working.  First off, the non-versioned filename appears to be some mishmash of shortcut and dll.  I'm not sure what's going on there, but trying to `dlopen` it doesn't work.

Secondly, the versioned filename is not loading properly.  Looking at it with DepWalker reveals a dependency on `cygwin1.dll`.  Is that......normal?

tkelman commented:
That means it used the wrong toolchain, it wasn't correctly cross compiled.

staticfloat committed:
We need to explicitly pass `CC` and `AR`, as they're not exported

staticfloat commented:
Alright, serious progress.  With [this binary](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-6e1377f2a4-win64.exe), I now get:
```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser-2.7.dll"))
"C:\\Users\\Elliot Saba\\AppData\\Local\\Julia-0.6.0-dev\\bin\\libhttp_parser-2.7.dll"
```

The only remaining problem is the weird `libhttp_parser.dll` file.  @tkelman what's SOP on windows; install only the versioned filename and use that explicitly within `dlopen()` calls?  Install both?  Install only unversioned filename?

tkelman commented:
If you put a version name in the dll, then should not use the unversioned name at all. Unless you're building for cygwin, which we aren't.

staticfloat commented on deps/http_parser.mk:
```diff
@@ -7,7 +7,7 @@ $(eval $(call git-external,http_parser,HTTPPARSER,,,$(BUILDDIR)))
 # person (@staticfloat) has the PLATFORM envvar set to something else.
 # Also manually set PREFIX to DESTDIR
-HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix)
+HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_prefix) LIBDIR=$(build_shlibdir)
```
My feeling is that they didn't really have windows in mind when they designed this.

tkelman replied:
if they're putting anything other than the dll there, thrn it's not right to put it in bin

staticfloat replied:
[They're not](https://github.com/nodejs/http-parser/blob/master/Makefile#L125-L128). 

Could `ln -s` be creating a nasty file on Windows?  Are we not allowed to have symlinks on Windows?  If not, my proposed solution is to just add an `rm libhttp_parser.dll` to the windows install steps and tell everyone to add the fully versioned name to any bindeps scripts that need it.

tkelman replied:
yes, if run in cygwin then ln -s creates a file that only cygwin can read. if run in msys, ln -s just copies i believe, so it's never really the right thing to do for mingw.

staticfloat committed:
On windows, move http_parser .dll files around a bit

staticfloat committed:
Nevermind, let's use the versioned filename

staticfloat commented on deps/http_parser.mk:
```diff
@@ -12,6 +12,7 @@ HTTPPARSER_FLAGS := PLATFORM=$(shell echo $(OS) | tr A-Z a-z) PREFIX=$(build_pre
 $(eval $(call staged-install, \
 	http_parser,$$(HTTPPARSER_SRC_DIR), \
 	MAKE_INSTALL,$(HTTPPARSER_FLAGS),, \
+	[[ "$(OS)" == "WINNT" ]] && mv $(build_shlibdir)/libhttp_parser-2.7.dll $(build_shlibdir)/libhttp_parser.dll; \
```
Ah, good to know.  Reverted.

staticfloat commented:
Looks like this last commit gets everything on the up-and-up.  Final windows binaries:

[Win x86](https://s3.amazonaws.com/julianightlies/bin/winnt/x86/0.6/julia-0.6.0-8fdea47fd9-win32.exe)
[Win x64](https://s3.amazonaws.com/julianightlies/bin/winnt/x64/0.6/julia-0.6.0-8fdea47fd9-win64.exe)

You have to access the library by its versioned filename:
```
julia> Libdl.dlpath(Libdl.dlopen("libhttp_parser-2.7.dll"))
"C:\\Users\\Elliot Saba\\AppData\\Local\\Julia-0.6.0-dev\\bin\\libhttp_parser-2.7.dll"
```

tkelman commented:
Or update the patch to be consistent with what got merged. Either way I realized this is of limited usefulness since not everyone will be building libgit2 this way, so like with mbedtls the package will need to have a fallback to deal with cases where the libgit2 that Julia uses is built differently.

staticfloat committed:
Sync libgit2 http-parser patch with upstream

staticfloat commented:
With the advent of things like HTTP.jl, I think this is less useful than it would have been a year ago, and I'm just going to close this for now.
